### DISCUSSION TITLE
Introducing mason.nvim

### DISCUSSION THREAD

**Discussion by williamboman** [williamboman/nvim-lsp-installer#876] [Announcements]
# Introducing mason.nvim

[`mason.nvim`](https://github.com/williamboman/mason.nvim) is the next generation version of `nvim-lsp-installer`. It builds on top of the very same foundation as `nvim-lsp-installer`, but with a majority of internals refactored to improve extensibility and testability.

More importantly, the scope of `mason.nvim` has also been widened to target more than just LSP servers. `mason.nvim` supports DAP servers, linters, formatters, and more. As of writing, `mason.nvim` provides 150+ packages for 100+ languages. It can be thought of as a general-purpose package manager, native to Neovim, that runs everywhere Neovim runs (Windows, macOS, Linux, etc.).

Another big change with `mason.nvim` is that executables are now linked to a single, shared, location, allowing seamless access from Neovim builtins (shell, terminal, etc.) as well as other 3rd party plugins.

There have also been other improvements such as broader test coverage, more documentation, and a revamped UI!

## What's the future of nvim-lsp-installer?

`nvim-lsp-installer` will no longer be actively maintained, meaning I won't be doing things like porting new mason.nvim packages to nvim-lsp-installer, monitoring lspconfig for changes that require updates, or put much effort into user support or bug issues (the recommendation for the latter will be to migrate to mason.nvim).

## [Why? <img src="https://camo.githubusercontent.com/49fbf4b3f9f1f0e2cf49961123b55200feba9dda271c6fdc90887d9acc79a564/68747470733a2f2f632e74656e6f722e636f6d2f4b6a4a544251396c66747341414141432f7768792d6875682e676966" width="30" />](https://camo.githubusercontent.com/49fbf4b3f9f1f0e2cf49961123b55200feba9dda271c6fdc90887d9acc79a564/68747470733a2f2f632e74656e6f722e636f6d2f4b6a4a544251396c66747341414141432f7768792d6875682e676966)

As more and more workflows and editing capabilities finds it way into the Neovim ecosystem (LSP, DAP, null-ls to name a few popular ones) there's an ever-increasing dependency on external tooling to be installed. While in many cases the process of installing this external tooling is simply a matter of a single command, discovering new tooling as well as managing your installed ones quickly gets complicated.

`mason.nvim` provides a single, Neovim native, interface for all of this, allowing you to get up and running in a matter of seconds. All packages are installed into a single location, allowing you to do things such as (i) easily backup your external dependencies, (ii) not require root to install something, and (iii) mount it as a volume in ephemeral Docker containers.

**Why not continue the work in nvim-lsp-installer?**
nvim-lsp-installer is very tightly coupled with lspconfig, increasing its scope would not only be awkward but it would very likely come with lots of breaking changes (something I've been careful not to introduce in the past). With more and more LSP-related capabilities continuously being added to Neovim core, setting up LSP servers without going through lspconfig is becoming more of an option. Oh also, the name of the plugin would also be very misleading :).

## Migrating from nvim-lsp-installer to mason.nvim

Migrating to `mason.nvim` should be rather easy. The following steps are a recommendation that highlights things to pay attention to:

1. Uninstall all of your installed servers. `mason.nvim` is not compatible with `nvim-lsp-installer`'s file hierarchy.
   Make sure to uninstall your existing servers to free up disk space. This can either be done manually, or via the
   `:LspUninstallAll` command.

2. Remove `nvim-lsp-installer` and add `mason.nvim` &
   [`mason-lspconfig.nvim`](https://github.com/williamboman/mason-lspconfig.nvim). If you use a plugin manager, update
   your configuration accordingly, example:

```diff
- use { "williamboman/nvim-lsp-installer" }
+ use { "williamboman/mason.nvim" }
+ use { "williamboman/mason-lspconfig.nvim" }
```

3. Set up `mason.nvim` in your Neovim init script. Refer to `:h mason-default-settings` & `:h
   mason-lspconfig-default-settings` for more details.

   Note that some settings have been renamed, for example `ui.icons.server_installed -> ui.icons.package_installed`.

   _If you still use the old, deprecated, `.on_server_ready()` API, [see this discussion first](https://github.com/williamboman/nvim-lsp-installer/discussions/636)._

```diff
- require("nvim-lsp-installer").setup {
-     ensure_installed = { "sumneko_lua" },
-     ui = {
-         icons = {
-             server_installed = "✓"
-         }
-     }
- }
+ require("mason").setup {
+     ui = {
+         icons = {
+             package_installed = "✓"
+         }
+     }
+ }
+ require("mason-lspconfig").setup {
+     ensure_installed = { "sumneko_lua" },
+ }
```

Other things worth paying attention to is that the package names in mason.nvim no longer follow the lspconfig naming scheme, but instead follows upstream naming schemes (so for example, `sumneko_lua` is now `lua-language-server`). The `mason-lspconfig` extension helps translate these names, more info in docs.

**Comment by ya2s** [comment]
I'm not happy with the more plugins I have to install, do I need [mason-lspconfig.nvim](https://github.com/williamboman/mason-lspconfig.nvim)?
Why did you split into two plugins?

**Reply by eeeXun** [reply]
Author mentions there https://github.com/williamboman/mason.nvim/pull/109

**Reply by vladdoster** [reply]
This comment makes no sense. Mason replaces `nvim-lsp-installer`, and has new features such as installing formatters and linters. 

The initial post states:

> [Why?]
> 
> As more and more workflows and editing capabilities finds it way into the Neovim ecosystem (LSP, DAP, null-ls to name a few popular ones) there's an ever-increasing dependency on external tooling to be installed. While in many cases the process of installing this external tooling is simply a matter of a single command, discovering new tooling as well as managing your installed ones quickly gets complicated.
> 
> mason.nvim provides a single, Neovim native, interface for all of this, allowing you to get up and running in a matter of seconds. All packages are installed into a single location, allowing you to do things such as (i) easily backup your external dependencies, (ii) not require root to install something, and (iii) mount it as a volume in ephemeral Docker containers.
> 
> Why not continue the work in nvim-lsp-installer?
> nvim-lsp-installer is very tightly coupled with lspconfig, increasing its scope would not only be awkward but it would very likely come with lots of breaking changes (something I've been careful not to introduce in the past). With more and more LSP-related capabilities continuously being added to Neovim core, setting up LSP servers without going through lspconfig is becoming more of an option. Oh also, the name of the plugin would also be very misleading :).

**Comment by TornaxO7** [comment]
Is this repo going to be archived?

**Reply by williamboman** [reply]
Nope!

**Reply by TornaxO7** [reply]
May I ask for the reason?

**Reply by williamboman** [reply]
When archiving a repo it becomes entirely frozen and I'd like to keep it open. This plugin still see a lot of usage and while I haven't committed to any support guarantees I'd say the current status resembles one that is somewhere between the traditional LTS and EOL policies. I'll be observing usage trends and if they're not trending in the right direction I'll gradually do more to nudge users to migrate.
